### PR TITLE
Evaluate the product basename based on `$BAZEL_OUTPUTS_PRODUCT`

### PR DESCRIPTION
**Pull request by gferon**
We discovered this problem when using a `select()` on the `bundle_name` of a `ios_application` where the condition comes from a `build_configuration` transition.

Without this patch, the build phase "Copy Bazel Outputs / Generate Bazel dependencies" gets a script with hardcoded value for the basename of the output, which actually _might_ vary dependending on the build configuration (when `bundle_name` or `executable_name` is dynamically evaluated).

This patch:
- Fixes the issue by using an environment variable available to the script.
- Adds three `config_setting` to actually make use of the `build_configuration` transition and use `select` with `bundle_name`.

We would greatly appreciate any feedback you may have. Thank you!

/cc @ohitsdaniel

### CODE REVIEW

File: examples/integration/BUILD
Comment (brentleyjones) [inline examples/integration/BUILD:42]: f

File: examples/integration/BUILD
Comment (brentleyjones) [inline examples/integration/BUILD:105]: Can you make this change in its own PR to land before this?

Comment (gferon) [reply examples/integration/BUILD:105]: Absolutely, I suppose maybe CI has to be changed as well?

**Review by brentleyjones — CHANGES_REQUESTED** [review]
Good find, and nice fix!

File: examples/integration/BUILD
Comment (brentleyjones) [reply examples/integration/BUILD:105]: No, it already does this for fixtures. the CI is blocked by you not being a regular contributor. I can unblock that.

**Comment by brentleyjones** [comment]
In the root of the repository, can you run `./test/update_all_fixtures.sh` and push those changes? That will adjust the test fixtures for the changes you made.

**Comment by gferon** [comment]
> In the root of the repository, can you run `./test/update_all_fixtures.sh` and push those changes? That will adjust the test fixtures for the changes you made.

Done, my MacBook ended up doing an _actual_ TestFlight to regenerate fixtures 😄

File: tools/generators/legacy/src/Generator/AddTargets.swift
Comment (brentleyjones) [inline tools/generators/legacy/src/Generator/AddTargets.swift:207]: We still need this conditional. I think instead of the full remove like I suggested (sorry) we can change `outputs.productBasename` to `outputs.hasProductOutput` (similar to the `hasSwiftOutputs`.

Comment (gferon) [reply tools/generators/legacy/src/Generator/AddTargets.swift:207]: I have a question though: isn't checking `output.productBasename` the same as checking whether `BAZEL_OUTPUTS_PRODUCT` is set? I guess there's a subtle difference but I couldn't find any.

Comment (brentleyjones) [reply tools/generators/legacy/src/Generator/AddTargets.swift:207]: Hmm. Yeah. I guess we can unconditionally run the script. It was probably an optimization we don't need.

Comment (gferon) [reply tools/generators/legacy/src/Generator/AddTargets.swift:207]: Ok, then this change should be good to go. Are any other changes needed?

Comment (brentleyjones) [reply tools/generators/legacy/src/Generator/AddTargets.swift:207]: Not that I know of. I need about 15 mins before I can start to review this again. It’s probably going to be merged as is though.

File: examples/cc/test/fixtures/bwb.xcodeproj/project.pbxproj
Comment (brentleyjones) [reply examples/cc/test/fixtures/bwb.xcodeproj/project.pbxproj:699]: Per other comment: this is fine. The check in the script prevents additional work. We might want to optimize more later, but this has so little overhead, it's probably fine.

**Review by brentleyjones — APPROVED** [review]

**Comment by brentleyjones** [comment]
Errors from CI:

```
ERROR: /Users/administrator/buildbuddy.noindex/remote_build/e8cccc0a-f136-43e0-83c4-0b60991cb5a9/repo-root/examples/integration/iOSApp/Source/BUILD:26:16: configurable attribute "bundle_name" in //iOSApp/Source:iOSApp doesn't match this configuration. Would a default condition help?
Conditions checked:
 //:build_aaaaaaa
 //:build_b
 //:build_c
```

And probably unrelated to your change, but we might have some non-determinism:

```diff
--- a/examples/integration/test/fixtures/bwb.xcodeproj/xcshareddata/xcschemes/iOSAppUnitTests_Scheme.xcscheme
+++ b/examples/integration/test/fixtures/bwb.xcodeproj/xcshareddata/xcschemes/iOSAppUnitTests_Scheme.xcscheme
@@ -109,7 +109,7 @@
             <BuildableReference
                BuildableIdentifier = "primary"
                BlueprintIdentifier = "1765D91CC168F25A5BC51603"
-               BuildableName = "iOSApp.C.app"
+               BuildableName = "iOSApp.AAAAAAA.app"
                BlueprintName = "iOSApp"
                ReferencedContainer = "container:../../../../../test/fixtures/bwb.xcodeproj">
             </BuildableReference>
```

Can you another PR that does the target based changes, so we iterate on those separately? And then have this PR just be the fix and updated fixes from the fix alone?

**Review by brentleyjones — APPROVED** [review]

File: tools/generators/legacy/src/Generator/AddTargets.swift
Comment (brentleyjones) [inline tools/generators/legacy/src/Generator/AddTargets.swift:215]: Because of this change, you'll still need to run `./test/update_all_fixtures.sh` again.

**Comment by brentleyjones** [comment]
> Hope you will squash this monster of a branch 👻 when mergin git.

Yeah, they are all squash commits on merge :+1:.

File: examples/integration/iOSApp/Source/BUILD
Comment (gferon) [reply examples/integration/iOSApp/Source/BUILD]: Ok, I _think_ it should be good now, sorry for the back and forth 😅

**Review by brentleyjones — APPROVED** [review]

**Comment by brentleyjones** [comment]
Thanks! Please open the PR with the project change as well now, minimally so we can look into why it was failing on CI.

**Pull request merged**
